File: stkpush.php
<?php
// INCLUDE THE ACCESS TOKEN FILE
include 'accessToken.php';
date_default_timezone_set('Africa/Nairobi');

// Safaricom API URLs and keys
$processrequestUrl = 'https://sandbox.safaricom.co.ke/mpesa/stkpush/v1/processrequest';
$callbackurl = 'https://1c95-105-161-14-223.ngrok-free.app/MPEsa-Daraja-Api/callback.php'; // Update with your callback URL
$passkey = "bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919";
$BusinessShortCode = '174379';
$Timestamp = date('YmdHis');

// ENCRYPT DATA TO GET PASSWORD
$Password = base64_encode($BusinessShortCode . $passkey . $Timestamp);

// Retrieve the money and phone values from the POST request
$phone = $_POST['phone'];
$money = $_POST['money'];

$PartyA = $phone;
$PartyB = $_POST['phone'];
$AccountReference = 'ERAMLINK NETWORKS';
$TransactionDesc = 'stkpush test';
$Amount = $money;

$stkpushheader = ['Content-Type:application/json', 'Authorization:Bearer ' . $access_token];

// INITIATE CURL
$curl = curl_init();
curl_setopt($curl, CURLOPT_URL, $processrequestUrl);
curl_setopt($curl, CURLOPT_HTTPHEADER, $stkpushheader); // Setting custom header
$curl_post_data = array(
  // Fill in the request parameters with valid values
  'BusinessShortCode' => $BusinessShortCode,
  'Password' => $Password,
  'Timestamp' => $Timestamp,
  'TransactionType' => 'CustomerPayBillOnline',
  'Amount' => $Amount,
  'PartyA' => $PartyA,
  'PartyB' => $BusinessShortCode,
  'PhoneNumber' => $PartyA,
  'CallBackURL' => $callbackurl,
  'AccountReference' => $AccountReference,
  'TransactionDesc' => $TransactionDesc
);

$data_string = json_encode($curl_post_data);
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
curl_setopt($curl, CURLOPT_POST, true);
curl_setopt($curl, CURLOPT_POSTFIELDS, $data_string);
$curl_response = curl_exec($curl);

// Debugging: Output the full response
echo $curl_response;

$data = json_decode($curl_response);
$CheckoutRequestID = $data->CheckoutRequestID;
$ResponseCode = $data->ResponseCode;

if ($ResponseCode == "0") {
  echo "The CheckoutRequestID for this transaction is: " . $CheckoutRequestID;
} else {
  echo "There was an error processing the request. Response: " . $curl_response;
}

curl_close($curl);

File: stkpush.py
import requests
from requests.auth import HTTPBasicAuth
import base64
import os
import json
from dotenv import load_dotenv
from datetime import datetime

# Load environment variables from .env file
load_dotenv()

# M-Pesa credentials
consumer_key = os.getenv("MPESA_CONSUMER_KEY")
consumer_secret = os.getenv("MPESA_CONSUMER_SECRET")
business_shortcode = os.getenv("MPESA_SHORTCODE")
lipa_na_mpesa_passkey = os.getenv("MPESA_PASSKEY")
callback_url = os.getenv("MPESA_CALLBACK_URL")

# Access token URL
access_token_url = 'https://sandbox.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials'

def get_access_token():
    try:
        response = requests.get(access_token_url, auth=HTTPBasicAuth(consumer_key, consumer_secret))
        response.raise_for_status()
        return response.json().get('access_token')
    except requests.exceptions.RequestException as e:
        print(f"Error obtaining access token: {e}")
        return None

def process_stkpush(amount, phone_number):
    access_token = get_access_token()
    if not access_token:
        return {"ResponseCode": "1", "ResponseDescription": "Failed to obtain access token"}

    # Generate timestamp
    timestamp = datetime.now().strftime('%Y%m%d%H%M%S')

    # Password
    password = base64.b64encode((business_shortcode + lipa_na_mpesa_passkey + timestamp).encode()).decode('utf-8')

    # Request data
    data = {
        'BusinessShortCode': business_shortcode,
        'Password': password,
        'Timestamp': timestamp,
        'TransactionType': 'CustomerPayBillOnline',
        'Amount': amount,
        'PartyA': phone_number,
        'PartyB': business_shortcode,
        'PhoneNumber': phone_number,
        'CallBackURL': callback_url,
        'AccountReference': 'BingwaSokoni',
        'TransactionDesc': 'Bingwa Sokoni payment'
    }

    # Lipa na Mpesa Online API endpoint
    stk_push_url = 'https://sandbox.safaricom.co.ke/mpesa/stkpush/v1/processrequest'

    # Headers
    headers = {
        'Authorization': f'Bearer {access_token}',
        'Content-Type': 'application/json'
    }

    try:
        # Send the request
        response = requests.post(stk_push_url, json=data, headers=headers)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error sending STK push request: {e}")
        return {"ResponseCode": "1", "ResponseDescription": "Failed to send STK push request"}

if __name__ == "__main__":
    # Example usage
    phone_number = input("Enter phone number: ")
    amount = input("Enter amount: ")
    response = process_stkpush(amount, phone_number)
    print(response)
